spring-shell-core/src/main/java/org/springframework/shell/command/CommandCatalog.java

```diff
@@ -0,0 +1,189 @@
+	interface CommandCatalogCustomizer {
+
+		/**
+		 * Customize a command catalog.
+		 *
+		 * @param commandCatalog a command catalog
+		 */
+		void customize(CommandCatalog commandCatalog);
+	}
+
+	/**
+	 * Gets an instance of a default {@link CommandCatalog}.
+	 *
+	 * @return default command catalog
+	 */
+	static CommandCatalog of() {
+		return new DefaultCommandCatalog(null, null);
+	}
+
+	/**
+	 * Gets an instance of a default {@link CommandCatalog}.
+	 *
+	 * @param resolvers the command resolvers
+	 * @param shellContext the shell context
+	 * @return default command catalog
+	 */
+	static CommandCatalog of(Collection<CommandResolver> resolvers, ShellContext shellContext) {
+		return new DefaultCommandCatalog(resolvers, shellContext);
+	}
+
+	/**
+	 * Default implementation of a {@link CommandCatalog}.
+	 */
+	static class DefaultCommandCatalog implements CommandCatalog {
+
+		private final HashMap<String, CommandRegistration> commandRegistrations = new HashMap<>();
+		private final Collection<CommandResolver> resolvers = new ArrayList<>();
+		private final ShellContext shellContext;
+
+		DefaultCommandCatalog(Collection<CommandResolver> resolvers, ShellContext shellContext) {
+			this.shellContext = shellContext;
+			if (resolvers != null) {
+				this.resolvers.addAll(resolvers);
+			}
+		}
+
+		@Override
+		public void register(CommandRegistration... registration) {
+			for (CommandRegistration r : registration) {
+				String commandName = commandName(r.getCommands());
+				commandRegistrations.put(commandName, r);
+			}
+		}
+
+		@Override
+		public void unregister(CommandRegistration... registration) {
+			for (CommandRegistration r : registration) {
+				String commandName = commandName(r.getCommands());
+				commandRegistrations.remove(commandName);
+			}
+		}
+
+		@Override
+		public void unregister(String... commandName) {
+			for (String n : commandName) {
+				commandRegistrations.remove(n);
+			}
+		}
+
+		@Override
+		public Map<String, CommandRegistration> getRegistrations() {
+			HashMap<String, CommandRegistration> regs = new HashMap<>();
+			regs.putAll(commandRegistrations);
+			for (CommandResolver resolver : resolvers) {
+				regs.putAll(resolver.resolve());
+			}
+			return regs.entrySet().stream()
+				.filter(e -> {
+					InteractionMode mim = e.getValue().getInteractionMode();
+					InteractionMode cim = shellContext != null ? shellContext.getInteractionMode() : InteractionMode.ALL;
+					if (mim == null || cim == null || mim == InteractionMode.ALL) {
+						return true;
+					}
+					else if (mim == InteractionMode.INTERACTIVE) {
+						return cim == InteractionMode.INTERACTIVE || cim == InteractionMode.ALL;
+					}
+					else if (mim == InteractionMode.NONINTERACTIVE) {
+						return cim == InteractionMode.NONINTERACTIVE || cim == InteractionMode.ALL;
+					}
+					return true;
+				})
+				.collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+		}
+
+		private static String commandName(String[] commands) {
```

Comment:
I think it would be an improvement to  encapsulate this into a method in CommandRegistration. The Builder talks about this detail and it would be great if that behavior was w/ the CR itself. 

```java
		 * Define commands this registration uses. Essentially defines a full set of
		 * main and sub commands. It doesn't matter if full command is defined in one
		 * string or multiple strings as "words" are splitted and trimmed with
		 * whitespaces. You will get result of {@code command subcommand1 subcommand2, ...}.
```

Reply:
Yes, and I think I'll polish interfaces around this.